File: stack/valid_parenthesis.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        if s[0] != '(' and s[0] != '{' and s[0] != '[':
            return False
        for i in range(len(s)):
            if s[i] == '(' or s[i] == '{' or s[i] == '[':
                stack.append(s[i])
            elif len(stack) >= 1:
                top = stack.pop()
                if s[i] == ')' and top != '(':
                    return False
                elif s[i] == '}' and top != '{':
                    return False
                elif s[i] == ']' and top != '[':
                    return False
            else:
                return False
        return len(stack) == 0

File: stack/valid_parenthesis.java
// Problem: https://leetcode.com/problems/valid-parentheses/
// Approach: Basic stack to hold most recently inputted parenthesis to be accessed LIFO
// Complexity: O(n) time, O(n) space


import java.util.Stack;

class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();

        if (s.charAt(0) == '(' && s.charAt(0) == '{' && s.charAt(0) == '[') {
            return false;
        } else {
            for (int i = 0; i < s.length(); i++) {
                char c = s.charAt(i);
                if (c == '(' || c == '{' || c == '[') {
                    stack.push(c);
                    continue;
                }

                if (c == ')' || c == '}' || c == ']') {
                    if (stack.isEmpty()) return false;
                    char top = stack.pop();
                    if (top == '(' && c != ')') {
                        return false;
                    } else if (top == '[' && c != ']') {
                        return false;
                    } else if (top == '{' && c != '}') {
                        return false;
                    }
                }
            }
        }
        return stack.isEmpty();
    }
}
